### DISCUSSION TITLE
Short circuit when a navigation only updates location state

### DISCUSSION THREAD

**Discussion by Synvox** [remix-run/react-router#10143] [Proposals]
Ran into an unexpected behavior today calling `navigate`:

```ts
function setValue(value) {
  navigate(
    location.pathname,
    {
      replace: true,
      state: {
        ...location.state,
        key: value
      },
    }
  )
}
```

Every time I called this function, my loaders would rerun. This makes sense because it's a navigation, but it doesn't actually navigate to a separate page. I wonder if this should short circuit like updating the location hash:

https://github.com/remix-run/react-router/blob/main/packages/router/router.ts#L1152

Thoughts?

**Comment by brophdawg11** [comment]
Related issue asking for the exact opposite behavior: https://github.com/remix-run/react-router/issues/12571 🙃 

Cross-posting my comment from there:

> As for the changing state, that's not currently exposed to loaders so they wouldn't return a different value anyway even if they did re-run. Any reason you can't use search params instead of history state? That would cause loaders to re-run and you'd have access to the new state in the loader.
>
> Even if we did expose state to loaders, it's not really enough to say "state change should trigger loaders" because here's an https://github.com/remix-run/react-router/discussions/10143 asking for the exact opposite behavior, so there's not going to be a one-size-fits all solution that works for everyone.
>
> I would recommend we use that existing proposal to discuss the possibility of adding state as a param to loaders/actions (and figuring out if this applies in a serializable way for SSR) and then you could opt in or out of running loaders on state change using a custom [dataStrategy](https://reactrouter.com/api/data-routers/createBrowserRouter#optsdatastrategy). Feel free to comment on your use case over there and if that gets enough traction it can be considered.

**Comment by brophdawg11** [comment]
We recently shipped an [`unstable_defaultShouldRevalidate`](https://api.reactrouter.com/v7/interfaces/react-router.NavigateOptions.html#unstable_defaultshouldrevalidate) API ([RFC](https://github.com/remix-run/react-router/discussions/10006)) that can be used on `Link`/`Form`/`useNavigate`/`useSubmit` to avoid revalidations in scenarios such as this.

**Reply by Synvox** [reply]
Oh sick. Thanks @brophdawg11!
